keps/sig-api-machinery/00xx-admission-webhooks-to-ga.md

```diff
@@ -0,0 +1,645 @@
+
+When deleting a collection, the webhook will get a filtered list of objects if the ObjectSelector is applied. The webhook may be called multiple times with sub-lists.
+
+The proposed change is to add an ObjectSelector to the webhook API both in v1 and v1beta1:
+
+```golang
+type Webhook struct {
+    ...
+     // ObjectSelector decides whether to run the webhook on an object based
+     // on whether the object.metadata.labels matches the selector. A namespace object must
+     // match both namespaceSelector (if present) and objectSelector (if present).
+     // For sub-resources, if the labels on the newObject is effective (e.g. pods/status)
+     // they will be used, otherwise the labels on the parent object will be used. If the
+     // sub-resource does not have a parent object, the call may fail or skipped based on
+     // the failure policy.
+     //
+     // Default to the empty LabelSelector, which matches everything.
+     // +optional
+     ObjectSelector *metav1.LabelSelector `json:"objectSelector,omitempty" protobuf:"bytes,7,opt,name=objectSelector"`
+}
+```
+
+### Scope
+
+Current webhook Rules applies to objects of all scopes. That means a Rule can uses wildcards to target both namespaced and cluster scoped objects. The proposal is to add a scope field to Admission Webhook configuration to limit webhook target on namespaced object or cluster scoped objects. This enables webhook developers to target all namespace objects or all cluster-scoped objects. The field will be added to both v1 and v1beta1. The field is optional and empty value means no scope restriction.
+
+```golang
+type ScopeType string
+
+const (
+     // ClusterScope means that scope is limited to cluster objects.
+     ClusterScope ScopeType = "Cluster"
+     // NamespacedScope means that scope is limited to namespaced objects.
+     NamespacedScope ScopeType = "Namespaced"
+)
+
+type Rule struct {
+    ...
+
+     // Scope specifies the scope of this rule. If unespecified, the scope is
+     // not limited.
+     //
+     // +optional
+     Scope ScopeType `json:"scope,omitempty" protobuf:"bytes,3,opt,name=scope"`
+}
+```
+
+### http for localhost
+
+Admission Webhook requires https for all connections. This is not necessary when the webhook in running on the same machine as apiserver (like a side-car or static pod). By relaxing this restriction, it would be possible to run side-cars or static pods on the same machine without cert management. The API change for this feature is documentation and relaxing validation.
```

Comment:
This is dangerous as traffic egresses in apiserver's own network space. We need to do something to ensure users cannot register webhooks which talk to random things colocated with apiserver. This will be a concern for any cloud provider that hosts Kubernetes.

The first, dumbest solution that comes to mind is a port whitelist in some on-disk configuration file. I'm sure you can think of something better.

Reply:
policy control is orthogonal to http/https, right? you can already register webhooks for `https://localhost:<port>`

Reply:
Arguably that isn't as much of a problem, IIRC you have to specify the serving cert, which you won't be able to get easily without access to the network.

HTTP opens up a new class of endpoints that previously were inaccessible.

Reply:
> HTTP opens up a new class of endpoints that previously were inaccessible.

Hmm... you're right. I thought there was a "skip TLS verification" option already, but that's on `APIService`, which only targets services. I suppose a local static webhook could bake in a "permanent" ca+cert unique to the webhook and pin the CA bundle in the webhook config. That would only allow calls to the webhook to connect to that webhook locally. It's ugly, but explicable. I'm ok with dropping this in favor of documentation.

Reply:
I like the idea of whitelisting some ports. We can even pass them as a parameter to API Server and if we have some in-binary webhooks, we can whitelist them in the code.